multicluster/controllers/multicluster/core/leader_elector.go

```diff
@@ -0,0 +1,126 @@
+/*
+Copyright 2021 Antrea Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package core
+
+import (
+	"context"
+	"crypto/rand"
+	"math/big"
+	"time"
+
+	"k8s.io/klog/v2"
+
+	"antrea.io/antrea/multicluster/controllers/multicluster/common"
+)
+
+/**
+ * Lets consider a ClusterSet with 2 leader clusters and 1 member cluster
+ *      Leader1        Leader2
+ *	     \         /
+ *	      \       /
+ *             -------
+ *               ^
+ *               |
+ *            Member1
+ * The member cluster connects to both leaders and performs Member Announce
+ * to announce itself. And all resources from the Members are written into
+ * Common Area of all Leader clusters. However resources are imported
+ * only from an elected-leader cluster.
+ *
+ * This interface performs the election to pick an elected-leader
+ * from the list of leader clusters in the ClusterSet spec.
+ *
+ * Leader Election picks an elected-leader among "connected" leaders
+ * by choosing one randomly
+ * The Member Cluster periodically writes MemberAnnounce CRD into every leader
+ * cluster and is considered "connected" if it can successfully write so.
+ * After leader election is done, the result is also used to update the
+ * spec of the MemberAnnounce CRD during the next periodic write into
+ * the leader's Common Area, so the leader knows it is the elected-leader
+ * of this member (mainly for visibility and reporting)
+ */
+type LeaderElector interface {
+	// start the leader elector, returns a cancelFunc to invoke when it must be stopped
+	StartLeaderElection() context.CancelFunc
```

Comment:
I do not see CancelFunc is used anywhere. Do we really need it?

Reply:
Yes I added it in the latest patch, it is needed to handle deletion of ClusterSet resource. But Ill remove the go routine from LeaderElector. Not sure I can remove all go routine, I think I still need one in RemoteCluster and in RemoteClusterManager. But Ill double check and let you know.